models/resnet/resnet_impl.hpp

```diff
@@ -0,0 +1,181 @@
+/**
+ * @file resnet_impl.hpp
+ * @author Aakash Kaushik
+ *
+ * Implementation of ResNet using mlpack.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MODELS_MODELS_RESNET_RESNET_IMPL_HPP
+#define MODELS_MODELS_RESNET_RESNET_IMPL_HPP
+
+#include "resnet.hpp"
+
+namespace mlpack {
+namespace models {
+
+template<typename OutputLayerType, typename InitializationRuleType,
+    size_t ResNetVersion>
+ResNet<OutputLayerType, InitializationRuleType, ResNetVersion>::ResNet() :
+    inputChannel(0),
+    inputWidth(0),
+    inputHeight(0),
+    numClasses(0)
+{
+  // Nothing to do here.
+}
+
+template<typename OutputLayerType, typename InitializationRuleType,
+    size_t ResNetVersion>
+ResNet<OutputLayerType, InitializationRuleType, ResNetVersion>::ResNet(
+    const size_t inputChannel,
+    const size_t inputWidth,
+    const size_t inputHeight,
+    const bool includeTop,
+    const bool preTrained,
+    const size_t numClasses) :
+    ResNet<OutputLayerType, InitializationRuleType, ResNetVersion>(
+        std::tuple<size_t, size_t, size_t>(
+        inputChannel,
+        inputWidth,
+        inputHeight),
+        includeTop,
+        preTrained,
+        numClasses)
+{
+  // Nothing to do here.
+}
+
+template<typename OutputLayerType, typename InitializationRuleType,
+    size_t ResNetVersion>
+ResNet<OutputLayerType, InitializationRuleType, ResNetVersion>::ResNet(
+    std::tuple<size_t, size_t, size_t> inputShape,
+    const bool includeTop,
+    const bool preTrained,
+    const size_t numClasses) :
+    inputChannel(std::get<0>(inputShape)),
+    inputWidth(std::get<1>(inputShape)),
+    inputHeight(std::get<2>(inputShape)),
+    numClasses(numClasses)
+{
+  // Config for different Versions.
+  if (ResNetVersion == 18)
+  {
+    numBlockArray = {2, 1, 2, 2};
+    builderBlock = "basicblock";
+  }
+  else if (ResNetVersion == 34)
+  {
+    numBlockArray = {3, 4, 6, 3};
+    builderBlock = "basicblock";
+  }
+  else if (ResNetVersion == 50)
+  {
+    numBlockArray = {3, 4, 6, 3};
+    builderBlock = "bottleneck";
+  }
+  else if (ResNetVersion == 101)
+  {
+    numBlockArray = {3, 4, 23, 3};
+    builderBlock = "bottleneck";
+  }
+  else if (ResNetVersion == 152)
+  {
+    numBlockArray = {3, 8, 36, 3};
+    builderBlock = "bottleneck";
+  }
+  else
+  {
```

Comment:
Since there are a lot of versions if-else statements look repetitive. Create a function that returns a map that whose key value pair is int - [vector,string]. If key exists set values, else throw an error.

Reply:
Hey, did this with maps, hope this is more elegant than the previou solution.